openmetrics/src/main/java/io/airlift/openmetrics/MetricsResource.java

```diff
@@ -0,0 +1,281 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+@Path("/v1/metrics")
+public class MetricsResource
+{
+    private static final Logger log = Logger.get(MetricsResource.class);
+    private static final String OPENMETRICS_CONTENT_TYPE = "application/openmetrics-text; version=1.0.0; charset=utf-8";
+    private static final String ATTRIBUTE_SEPARATOR = "_ATTRIBUTE_";
+    private static final String TYPE_SEPARATOR = "_TYPE_";
+    private static final String NAME_SEPARATOR = "_NAME_";
+    private static final Pattern METRIC_NAME_PATTERN = Pattern.compile("[[a-zA-Z]][\\w_]*");
+    private static final LoadingCache<String, Boolean> VALID_NAMES_CACHE = CacheBuilder.newBuilder().build(new CacheLoader<String, Boolean>()
+    {
+        @Override
+        public Boolean load(String metricName)
+        {
+            return METRIC_NAME_PATTERN.matcher(metricName).matches();
+        }
+    });
+
+    private final MBeanServer mbeanServer;
+    private final boolean inferNumeric;
+    private final boolean inferString;
+    private final boolean inferAll;
+
+    private final List<ObjectName> allMetricsObjectNames;
+
+    @Inject
+    public MetricsResource(MBeanServer mbeanServer, MetricsConfig metricsConfig)
+    {
+        this.mbeanServer = requireNonNull(mbeanServer, "mbeanServer is null");
+        this.inferNumeric = metricsConfig.isInferNumeric();
+        this.inferString = metricsConfig.isInferStrings();
+        this.inferAll = metricsConfig.isInferAll();
+
+        if (metricsConfig.getJmxDomains().isEmpty()) {
+            this.allMetricsObjectNames = ImmutableList.of(ObjectName.WILDCARD);
+        }
+        else {
+            this.allMetricsObjectNames = metricsConfig.getJmxDomains().stream().map(domain -> {
+                try {
+                    return new ObjectName(domain + ":*");
+                }
+                catch (MalformedObjectNameException e) {
+                    throw new IllegalArgumentException("Unable to parse domain : %s".formatted(domain), e);
+                }
+            }).collect(toImmutableList());
+        }
+    }
+
+    @GET
+    @Produces(OPENMETRICS_CONTENT_TYPE)
+    public String getMetrics(@QueryParam("name[]") List<String> filter)
+    {
+        StringBuilder body = new StringBuilder();
+        if (filter != null && !filter.isEmpty()) {
+            for (String metricName : filter) {
+                body.append(toMetricExposition(metricName));
+            }
+        }
+        else {
+            for (ObjectName metricObjectNames : allMetricsObjectNames) {
+                body.append(allMetricExpositions(metricObjectNames));
+            }
+        }
+        body.append("# EOF\n");
+        return body.toString();
+    }
+
+    private Set<ObjectName> objectNamesFromMetricName(String metricName)
```

Comment:
those domain conversion methods are tricky. 
I strongly suggest extracting those as a helper and cover with unit tests.